hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/publishing/PublishingHandler.java

```diff
@@ -0,0 +1,220 @@
+                messageContent -> processAndLogPreview(exchange, topicWithMetrics, state, messageTimeout, messageContent),
+                exception -> {
+                    messageTimeout.remove();
+                    if (exception instanceof InvalidContentLengthException) {
+                        messageErrorProcessor.invalidLength(exchange, topicWithMetrics.getTopic(), state.messageId(), exception);
+                    } else {
+                        messageErrorProcessor.readException(exchange, topicWithMetrics.getTopic(), state.messageId(), exception);
+                    }
+                });
+    }
+
+    private void process(HttpServerExchange exchange, TopicWithMetrics topicWithMetrics, MessageState state,
+                         MessageTimeout messageTimeout, byte[] messageContent) {
+
+        Message message;
+        try {
+            message = messageFactory.create(exchange.getRequestHeaders(), topicWithMetrics, state.messageId(), messageContent);
+            messageTimeout.setMessage(message);
+        } catch (InvalidMessageException | AvroConversionException | UnsupportedContentTypeException exception) {
+            messageTimeout.remove();
+            messageErrorProcessor.invalidMessage(exchange, topicWithMetrics.getTopic(), state.messageId(), exception);
+            return;
+        } catch (CouldNotLoadSchemaException | SchemaMissingException exception) {
+            messageTimeout.remove();
+            messageErrorProcessor.missingSchema(exchange, topicWithMetrics.getTopic(), state.messageId(), exception);
+            return;
+        }
+
+        // change state of exchange to dispatched,
+        // thanks this default response with 200 status code is not returned after handler finish its execution
+        exchange.dispatch();
+
+        state.set(SENDING_TO_KAFKA_PRODUCER_QUEUE);
+        StartedTimersPair brokerLatencyTimers = topicWithMetrics.startBrokerLatencyTimers();
+        brokerMessageProducer.send(message, topicWithMetrics.getTopic(), new PublishingCallback() {
+            @Override
+            public void onUnpublished(Message message, Topic topic, Exception exception) {
+                if (exception instanceof MissedProducerBufferException) {
+                    // in worker thread
+                    state.set(ERROR_IN_SENDING_TO_KAFKA_PRODUCER_QUEUE);
+                    brokerLatencyTimers.close();
+                    messageTimeout.remove();
+                    messageErrorProcessor.missedProducerBuffer(exchange, topic, state.messageId(), exception);
+                } else {
+                    // in kafka producer thread
+                    exchange.dispatch(() -> {
+                        brokerLatencyTimers.close();
+                        messageTimeout.remove();
+                        logger.warn("Exception while sending message to a broker. Message already sent to a producer buffer.", exception);
```

Comment:
fixed, if any exception is thrown from kafka producer then message is not in the buffer. Thanks for your help @wojtkiewicz with establishing this.